backend/database.js

```diff
@@ -51,6 +51,52 @@ export default (sbp('sbp/selectors/register', {
       }
     })
   },
+  'backend/db/streamEntriesBefore': async function (contractID: string, before: string, limit: number): Promise<*> {
+    let currentHEAD = await sbp('chelonia/db/latestHash', contractID)
+    if (!currentHEAD) {
+      throw Boom.notFound(`contractID ${contractID} doesn't exist!`)
+    }
+
+    const hash = before || currentHEAD
+    let prefix = '['
+    if (hash === currentHEAD) {
+      const entry = await sbp('chelonia/db/getEntry', currentHEAD)
+      prefix = `["${strToB64(entry.serialize())}",`
+      limit--
+    }
+    let metBefore = false
+
+    return new Readable({
+      async read (): any {
+        try {
+          const entry = await sbp('chelonia/db/getEntry', currentHEAD)
+          const json = `"${strToB64(entry.serialize())}"`
+
+          if (currentHEAD !== hash && metBefore) {
+            this.push(prefix + json)
+            prefix = ','
+            limit--
+          } else if (currentHEAD === hash) {
+            metBefore = true
+          }
+
+          if (!limit || currentHEAD === contractID) {
+            this.push(']')
+            this.push(null)
+          } else {
+            currentHEAD = entry.message().previousHEAD
+            this.push('')
```

Comment:
I think this code could be greatly simplified. I do not see a need for the `metBefore` variable.

Something like:

```js
  'backend/db/streamEntriesBefore': async function (before: string, limit: number): Promise<*> {
    let prefix = '['
    let currentHEAD = before
    let entry = await sbp('chelonia/db/getEntry', currentHEAD)
    if (!entry) {
      throw Boom.notFound(`entry ${currentHEAD} doesn't exist!`)
    }
    currentHEAD = entry.message().previousHEAD
    return new Readable({
      async read (): any {
        try {
          if (!currentHEAD || limit === 0) {
            this.push(']')
            this.push(null)
          } else {
            entry = await sbp('chelonia/db/getEntry', currentHEAD)
            const json = `"${strToB64(entry.serialize())}"`
            currentHEAD = entry.message().previousHEAD
            limit -= 1
            this.push(prefix + json)
            prefix = ','
          }
        } catch (e) {
          // TODO: properly return an error - I don't think this lets the caller know there was an error
          console.error(`read(): ${e.message}:`, e)
          this.push(']')
          this.push(null)
        }
      }
    })
  },
```

Notice:

- We don't need the `contractID`, so it's removed from the parameters
- We don't need `metBefore`
- Both `'backend/db/streamEntriesBefore'` and `'backend/db/streamEntriesSince'` fail to properly inform the caller of an error during the string. That's why I added the `TODO:` there. Maybe either you or @snowteamer know how to handle that properly? I'd check the Node.js documentation for this (and maybe the Hapi.js documentation)

Reply:
According to the [docs](https://nodejs.org/api/stream.html#errors-while-reading), if I understand them correctly no error must be manually thrown here in `read()`, but [destroy(error)](https://nodejs.org/api/stream.html#readabledestroyerror) must be used instead. This will result in an `error` event to be emitted, with the given error argument as payload.

Also note that `destroy()` calls `_destroy()` internally, which can be [overriden](https://nodejs.org/api/stream.html#readable_destroyerr-callback) in case custom behavior is needed.

Reply:
Improved and optimized this `streamEntriesBefore` sbp function
- removed `metBefore` variable indeed
- I didn't remove the `contractID` because there could be the cases when we are going to get 40 latest events. In this case, we need to know whether the event is the latest one or not. Also in this case, there is one thing different which is to set the default `currentHEAD`.

> Both 'backend/db/streamEntriesBefore' and 'backend/db/streamEntriesSince' fail to properly inform the caller of an error during the string. 

I have no idea right now when the errors occurs and how we should handle. I need to study this more.

Reply:
> - I didn't remove the `contractID` because there could be the cases when we are going to get 40 latest events. In this case, we need to know whether the event is the latest one or not. Also in this case, there is one thing different which is to set the default `currentHEAD`.

@Silver-IT But the problem with this approach is that it unnecessarily complicates the API and creates a somewhat unclean interface.

For example, with the way the code is currently written, it's possible to call the API with a `before` value from `contractA`, and a `contractID` value that corresponds to `contractB`. It will still work, but no error will be thrown, and this sort of API is therefore confusing. I don't think the answer is to throw an error in this instance - that will only complicate the code unnecessarily even more.

We want the Chelonia API to be a **simple** as possible - with no necessary complexities. From our CONTRIBUTING guidelines:

> A PR SHOULD be a minimal and accurate answer to exactly one identified and agreed problem.

If something can be simplified, it should be, so that it's **minimal** answer to the problem.

In Chelonia, events are a linked list pointing backwards in time:

```
[ ContractA ] <- [ Event1 ] <- [ Event2 ]  <- [ Event3 ]
```

The way this API is written right now asks for 3 inputs:

```
[ ContractA ] <- [ Event1 ] <- [ Event2 ]  <- [ Event3 ]
     ^                                        ^
     |                                        |
Input1                                        Input2

40 <--- Input3
```

But a simpler, more minimal API exists that only asks for 2 inputs:

```
[ ContractA ] <- [ Event1 ] <- [ Event2 ]  <- [ Event3 ]
                                              ^
                                              |
                                            Input1

40 <--- Input2
```

If you want the 40 latest events, then before calling this API, you can call `'chelonia/private/out/latestHash'`.

In fact, we should move that to be a public API (move it from `internals.js` to `chelonia.js`), and rename it to `'chelonia/out/latestHash'`. And then call `'chelonia/out/eventsSince'`.

Reply:
> For example, with the way the code is currently written, it's possible to call the API with a before value from contractA, and a contractID value that corresponds to contractB. It will still work, but no error will be thrown, and this sort of API is therefore confusing. I don't think the answer is to throw an error in this instance - that will only complicate the code unnecessarily even more.

Make sense.

One more thing is that the reason why we need to know is that
If the `before` is the latest one, we need to get 40 events from `before` including **itself**.
If the `before` is not the latest one, we need to get 40 events before `before` hash.

Reply:
Just making a note here to record that based on Slack conversation we figured out the solution to this would be to simply always include `before`

Reply:
Perfect.
So the return value should contain events like the following.
up to `limit` number of events before `before` + `before`